src/rw.jl

```diff
@@ -0,0 +1,196 @@
+using ADIOS2
+
+    """
+    Intialize ADIOS2. Supports several configuration options: MPI, serial,
+    existing XML config file or a new file with specified parameters.
+    """
+    function adios2_init(; filename::AbstractString = "", engine_type = "",
+                         serial = false, comm = nothing)
+
+        use_mpi = !isnothing(comm)
+
+        adios = nothing
+
+        if isempty(filename)
+            if serial
+                global adios = adios_init_serial()
+            elseif use_mpi
+                filename = adios2_config(engine = engine_type)
+                global adios = ADIOS2.adios_init_mpi(joinpath(pwd(), filename), comm)
+            else serial_file
+                filename = adios2_config(engine = engine_type)
+                global adios = ADIOS2.adios_init_serial(joinpath(pwd(), filename))
+            end
+        else
+            if use_mpi
+                global adios = ADIOS2.adios_init_mpi(joinpath(pwd(), filename), comm)
+            else
+                global adios = ADIOS2.adios_init_serial(joinpath(pwd(), filename))
+            end
+        end
+
+        global init_state = true          # mark the initialization step
+        global vars = []                  # initialize the variable array (used later for reading and writing)
+        global nprocessed = 0            # needed for the reading step
+        global reading_now = false        # flag for the reading loop (to avoid race conditions)
+    end
+
+    """
+    Initialize io in read mode and the corresponding engine for reading data.
+    """
+    function read_setup(bp_filename = "")
+
+        io = ADIOS2.declare_io(adios, "readerIO")
+        bp_path = joinpath(pwd(), bp_filename)
+        global engine = ADIOS2.open(io, bp_path, mode_read)    # Open the file/stream from the .bp file
+
+    end
+
+    """
+    Initialize io in write mode and the corresponding engine for writing data.
+    """
+    function write_setup(variables ...; bp_filename = "")
+
+        io = ADIOS2.declare_io(adios, "IO")
+
+        for var_pair in variables
+            var_id = define_variable(io, var_pair.first, eltype(var_pair.second))  # Define a new variable
+            push!(vars, var_id)
+        end
+
+        bp_path = joinpath(pwd(), bp_filename)
+        global engine = ADIOS2.open(io, bp_path, mode_write)   # Open the file/stream from the .bp file
+
+    end
+
+    """
+    Perform the update using specified variables.
+    """
+    function write(update_var = nothing ...)
+
+        for v in vars
+            begin_step(engine)                                       # Begin ADIOS2 write step
```

Comment:
`begin_step` needs to be before the loop and `end_step` after the loop. All variables should be written in one step.